object-model/copy-semantics.md

```diff
@@ -0,0 +1,151 @@
+# C++ object model: copy semantics
+_Skeleton instructions are typeset in italic text._
+
+## Overview
+
+_Provides a short natural language abstract of the module’s contents._
+_Specifies the different levels of teaching._
+
+<table>
+  <thead>
+    <th>Level</th>
+    <th>Objectives</th>
+  </thead>
+  <tr>
+    <td>Foundational knowledge</td>
+    <td>understanding how and when are copies made</td>
+  </tr>
+  <tr>
+    <td>Advanced</td>
+    <td>implementing user-defined copy operations</td>
+  </tr>
+  <tr>
+    <td>Further studies</td>
+    <td>special cases: copy elision</td>
+  </tr>
+</table>
+
+## Motivation
+
+_Why is this important?_
+_Why do we want to learn/teach this topic?_
+
+Copy semantics allows the user to determine how objects of a class get replicated and interact on a value level.
+
+## Topic introduction
+
+_Very brief introduction to the topic._
+
+Explains when and how objects are copied.
+
+## Foundational knowledge: How and when are copies made
+
+### Background/Required Knowledge
+
+A student:
+* knows what a type is?             [[C++ object model: types]][1]
+* knows what an object is?          [[C++ object model: objects]][2], [[C++ object model: constant objects]][3]
+* knows what class invariants are?
+
+It helps when a student:
+* knows move semantics              [[C++ object model: move semantics]][4]
+* knows special member functions    [[C++ object model: special member functions]][5]
+
+### Student outcomes
+
+_A list of things “a student should be able to” after the curriculum._
+_The next word should be an action word and testable in an exam._
+_Max 5 items._
+
+A student should be able to:
+
+1. explain what copy semantics accomplish
+  * establishing “equivalent” object state in another object
+2. explain difference between deep and shallow copy
+3. explain where copies are made
+
+### Caveats
+
+_This section mentions subtle points to understand, like anything resulting in
+implementation-defined, unspecified, or undefined behavior._
+
+* Compiler-provided (shallow) copies may result in ownership problems (e.g. `char*`).
```

Comment:
My verbose two cents on lines 64 and 72 and 82–85, FWIW:

I personally try not to use the terms "deep copy" and "shallow copy" — or mention the terms only so that I can deflate them. These terms make sense in Java or Python, where lots of things are secretly pointers under the hood; e.g.
```
a = [1,2,3]
b = a  # "shallow copy"
a[0] += 10
assert b == [11,2,3]  # surprise! b was not a copy of a, after all
```
But in C and C++, all pointers are explicit; so we don't have two different kinds of copy. We have a single privileged operation that we simply call "copy":
```
std::string *x = &s;
std::string *y = x;  // "copy" the std::string*'s value (but not what it points _to_)
std::string a = "hello";
std::string b = a;  // "copy" the std::string's value
std::string_view a = "hello";
std::string_view b = a;  // "copy" the std::string_view's value
```
If you write a `my::string` class where when you "make a copy" you end up with the wrong thing (two objects with unequal values, or one or both of which have broken their class invariants), then you don't have a "shallow copy" — you have a "buggy copy." On the other hand, if you implement your copy operations correctly, then you don't have either "shallow copy" or "deep copy"; all you have is "copy."

If I have `struct S { char *p; S(const S&) { ??? } }`, I do not know what to put in the `???` unless I know what are my class invariants. My class invariants dictate how I must implement copy. (Not "deep copy." Just, "copy.")

However, note that my teaching experience is exclusively with people who are already software engineers, so it's easy for me to appeal to their intuition: "If your copy constructor does what you intended, then it's good. If it doesn't do the right thing, don't dignify it with a name like 'shallow'; it's just buggy, that's all." This _might_ be less effective on people with less experience translating their intentions into code.

Reply:
Hi, sorry for the late reply, I was kinda bussy in the last weeks.

Regarding deep/shallow copies:
We discuss this in our SG20-group at the Prague meeting in detail. Our reasoning was that these two terms refer to semantic concepts a programmer has to understand when implementing copy operations. This relates to what you mean by "buggy copy" and thinking about class invariants. When implementing, for example, a copy constructor we need to reason about: Do we want to make this a `deep copy` or should we make it a `shallow` one. Your correct, what you need is dictated by the class invariants but how we differentiate the cases is what we refer to as `deep` vs `shallow`. Do my invariants need an independent object of type `T` or can I reuse the old one?
Especially important here is that `deep copy` implies transitivity, so I need to make sure if all the copies of my members and their members "to the right thing", regarding my invariants. Like in a multi threaded case where I want to prevent concurrent accesses.

For example, if the class contains a `std::shared_ptr`.
```cpp
class Foo {
public:
  Foo(const &Other) {
    ...
  }
private:
  std::shared_ptr<Bar> member_bar;
};
```

Maybe, a `shallow copy` is sufficient and I can just copy `member_bar` or I need to create my own `Bar`, i.e., copy bar and create a new `shared_ptr`. There could also be cases where both are required and the `copy operation` is implemented shallow but an extra member function does a `deep copy`.

All in all, a student should be able to differentiate the two semantically different cases and reason about the trade offs.

Reply:
> When implementing, for example, a copy constructor we need to reason about: Do we want to make this a deep copy or should we make it a shallow one. You're correct, what you need is dictated by the class invariants but how we differentiate the cases is what we refer to as deep vs shallow. Do my invariants need an independent object of type T or can I reuse the old one?

I agree with your last sentence, but continue to disagree with the "deep/shallow" distinction. To paraphrase your last sentence: If my object depends on some other object of type T, do I need to _copy_ that T object, or _not?_

> For example, if the class contains a `std::shared_ptr`...

You're doing class design backwards. In real life you don't _start_ with a list of members and then think about class semantics; you start with class semantics and then _from those semantics_ you decide what your members should look like.

For example, in one codebase I have a `Transaction` class. A `Transaction t` has methods `t.rule_matched()`, which tells which user-configured `Rule` object applied to this transaction; and `t.protocol()`, which returns either `"tcp"` or `"udp"`. To make a `Transaction` copyable, I'd have to think about: Should _copying_ a transaction cause the _copying_ of the protocol string? (Kinda sorta. Certainly I could use a `std::string` data member for that. But I could also get away with simply an `enum Protocol` data member.) Should _copying_ a transaction cause a _copying_ of the `Rule` that applied to it? (Definitely not. The `Transaction` object does not _contain_ the rule that applied to it; it merely _refers_ to the rule. So I should use a `Rule*` or perhaps a `std::shared_ptr<Rule>` to express that _refers-to_ relationship. Either way, _copying_ the pointer represents copying the _refers-to_ relationship, which matches my intention.)

Everything falls out naturally — in fact, it all follows the Rule of Zero! — and I never had to use the words "deep" or "shallow." I only had to talk about "copying" or "not copying." I never had to talk about _how_ to copy, because in C++ there is only one way to copy — you call _the_ copy constructor.

Does that clarify my position?

Reply:
To amplify I’d like to quote Tony van Eerd and me:

> Copy or copy not, there is no shallow!

Do not teach the mistakes of bad design taken from other OO languages and apply it to C++. I suffered through that in the early 1990s with the NIHCL, do not teach those designs!

Reply:
I understand your position. What I wanted to remark is that we chose `deep` and `shallow` as a mental model to think about what copy means. Not that there exist two different 'modes' in general in C++.

Referring to your example, if you have a _refers-to_ relationship would mean a `shallow copy` in general, where copying the `Rule` object would be `deep`.

Overall, I'm personally not to tied to the terms used to express things here and if we want to name the problem description in the guide differently I'm fine with it. But we probably should discuss this in the next SG20 meeting, which is next week Friday. @Quuxplusone join us if you like :grinning:, otherwise, if you cannot find time I'll try to present your idea/remark to the group.

Next meeting: Friday, May 8⋅17:00 – 18:00 CET (GMT+02:00)

@jcvw could you please put this on the agenda for me: "Should we refer to deep/shallow copy in the copy semantics topic."

Reply:
@Quuxplusone I reworded the sections, as discussed on Friday. Any suggestions?